presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -657,5 +883,96 @@ private List<Subfield> findSubfields(String rootName)
                         .collect(toImmutableList());
             }
         }
+
+        /**
+         * Represents the description of the array function and contains the properties important for subfield pruning.
+         */
+        private static final class ArrayFunctionDescriptor
+        {
+            /**
+             * Indicates whether the function changes the shape of the array element.
+             * If function does not change the shape of the array element then it could safely be nested.
+             */
+            boolean isChangingElementShape;
+
+            /**
+             * Indicates whether the function outputs all subfields.
+             * If function does not output all subfield then we can prune all subfields that were not accessed by this function or nested functions.
+             */
+            boolean isOutputtingAllSubfields;
+
+            /**
+             * Index of the argument in CallExpression, which represents this function, that contains the functional parameter.
+             * Value -1 indicates that the function does not accept the functional argument.
+             */
+            int functionArgumentIndex;
+
+            /**
+             * The mapping of the index of lambda argument to the index of the array argument of the CallExpression, which represents this function.
+             */
+            Map<Integer, Integer> lambdaArgumentToArrayArgumentIndexMap;
+
+            /**
+             * List of indices of the array arguments in the CallExpression, which represents this function.
+             */
+            List<Integer> arrayArgumentIndexes;
+
+            /**
+             * Contains the functional parameter that needs to be applied to the pathPrefix.
+             */
+            Function<List<PathElement>, List<PathElement>> pathPrefixTransform;
```

Comment:
Here, the pathPrefix is the prefix that needs to be added to the accessed subfield. For instance, consider the query 
```
SELECT ANY_MATCH(array_column, x -> x.sub_filed > 0) FROM some_table
```
when traversing the expression, we capture the subfield as `x.sub_field`. `x` is only used to correctly identify from what array this sufield was accessed. Now, we have element's path `sub_field`. However, since `array_column` is of ArrayType, the correct path should be `[*].sub_field`. Therefore, *pathPrefix* `[*]` needs to be applied in this case to all subfields. Let's consider another exabmple
```
SELECT ANY_MATCH(FLATTEN(array_of_array_column), x -> x.sub_filed > 0) FROM some_table
```
Here, the path for the accessed `sub_filed` is `[*].[*].sub_filed` and therefore *pathPrefix* `[*].[*]` needs to be applied to the `sub_field` captured in the lambda expression.